src/java.base/share/classes/jdk/internal/foreign/abi/UpcallStubs.java

```diff
@@ -50,13 +50,13 @@ private static void freeUpcallStub(long stubAddress) {
         registerNatives();
     }
 
-    static MemorySegment makeUpcall(long entry, Arena scope) {
-        MemorySessionImpl.toMemorySession(scope).addOrCleanupIfFail(new MemorySessionImpl.ResourceList.ResourceCleanup() {
+    static MemorySegment makeUpcall(long entry, Arena arena) {
+        MemorySessionImpl.toMemorySession(arena).addOrCleanupIfFail(new MemorySessionImpl.ResourceList.ResourceCleanup() {
             @Override
             public void cleanup() {
                 freeUpcallStub(entry);
             }
         });
-        return MemorySegment.ofAddress(entry, 0, scope);
+        return MemorySegment.ofAddress(entry).reinterpret(arena.scope(), null);
```

Comment:
In principle we could just express this using public API if we tweak the semantics for cleanup action to always run in case the scope is found to be already closed.

Reply:
Yeah, I think that might be better as semantics in general. If the attach fails, don't leak, and run the cleanup action.

Reply:
Not super sure how to specify this though - e.g. should we say that the cleanup is always run when the method completes abnormally? Possible failure causes are:
* the scope is no longer alive (this is what we care about)
* the size is negative
* the method is called on a segment that is not native
* the method is called and restricted access is not enabled.

Note 100% sure *when* we should say we would attempt to run the cleanup action. Kind of tempted leave it as is.

Reply:
Ok, probably better to leave it as is for now.